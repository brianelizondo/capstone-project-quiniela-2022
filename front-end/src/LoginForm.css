.LoginForm-form-card{
    padding: 30px 30px;
    background-color: rgba(0, 0, 0, .35);
    border-radius: 10px !important;
}

.LoginForm-form-container button{
    font-size: 20px;
    font-weight: 700;
    text-transform: uppercase;
}