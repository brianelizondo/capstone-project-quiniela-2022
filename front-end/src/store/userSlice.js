import { createSlice } from '@reduxjs/toolkit';

export const userSlice = createSlice({
    name: 'user',
    initialState: {
        id: null, 
        firstName: null, 
        lastName: null, 
        email: null, 
        username: null, 
        isAdmin: false,
        token: null,
        isAuthenticated: false
    },
    reducers: {
        login: (state, action) => {
            state.id = action.payload.user.id;
            state.firstName = action.payload.user.firstName;
            state.lastName = action.payload.user.lastName;
            state.email = action.payload.user.email;
            state.username = action.payload.user.username;
            state.isAdmin = action.payload.user.isAdmin;
            state.token = action.payload.token;
            state.isAuthenticated = true;
        },
        update: (state, action) => {
            state = action.payload
        },
        logout: (state) => {
            state.id = null;
            state.firstName = null;
            state.lastName = null;
            state.email = null;
            state.username = null;
            state.isAdmin = null;
            state.token = null;
            state.isAuthenticated = false;
        }
    }
})
  
// Action creators are generated for each case reducer function
export const { login, update, logout } = userSlice.actions;
export default userSlice.reducer;