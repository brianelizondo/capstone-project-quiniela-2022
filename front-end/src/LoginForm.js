import React, { useState } from 'react';
import { useHistory } from "react-router-dom";
import { Form, FloatingLabel, Button, Spinner, Col, Row, Alert, Card } from "react-bootstrap";
import { useFormik } from 'formik';
import * as Yup from 'yup';
import './LoginForm.css';

import { useDispatch } from 'react-redux';
import { login } from './store/userSlice';

function LoginForm({ userAuthenticate }){
    // history, loading and state for submit button
    const history = useHistory();
    const [btnSubmitLoading, setBtnSubmitLoading] = useState(false);
    const [checkUserPass, setCheckUserPass] = useState(false);
    // redux selector and dispatch
    const dispatch = useDispatch();

    // config for user register form
    const formik = useFormik({
        initialValues: {
            username: "",
            password: ""
        },
        validationSchema: Yup.object({
            username: Yup.string()
                .required('Required')
                .matches(/^([a-zA-Z0-9])*$/, { message: "Only letters and numbers are allowed" })
                .min(8, 'Must be 8 characters or more')
                .max(20, 'Must be 20 characters or less'),
            password: Yup.string()
                .required('Required')
                .matches(/^([a-zA-Z0-9])*$/, { message: "Only letters and numbers are allowed" })
                .min(8, 'Must be 8 characters or more')
                .max(20, 'Must be 20 characters or less')
          }),
        onSubmit: async (values) => {
            setBtnSubmitLoading(true);
            const respAuthenticate = await userAuthenticate(values);

            if(respAuthenticate[0]){
                setBtnSubmitLoading(false);
                setCheckUserPass(true);
            }else{
                dispatch(login(respAuthenticate));
                history.push(`/users/${respAuthenticate.user.username}/profile`);
            }
        }
    });

    return (
        <div className="LoginForm col-md-8 offset-md-2">
            <Row>
                <Col>
                    <h1 className='section-title'>Login</h1>
                </Col>
            </Row>
            <Row>
                <Col>
                    <h5 className='section-subtitle'>Login to Quinielas World Cup 2022</h5>
                </Col>
            </Row>

            <div className="LoginForm-form-container col-md-8 offset-md-2">
                <Card className="LoginForm-form-card">
                    <Form onSubmit={formik.handleSubmit}>
                        <Row>
                            <Col>
                                <FloatingLabel controlId="username" label="Username" className="mb-3">
                                    <Form.Control name="username" type="text" placeholder="Username"
                                        onChange={formik.handleChange}
                                        onBlur={formik.handleBlur}
                                        value={formik.values.username}
                                        isInvalid={formik.touched.username && !!formik.errors.username}
                                    />
                                    <Form.Control.Feedback type="invalid">{formik.errors.username}</Form.Control.Feedback>
                                </FloatingLabel>
                            </Col>
                        </Row>
                        <Row>
                            <Col>
                                <FloatingLabel controlId="password" label="Password" className="mb-3">
                                    <Form.Control name="password" type="password" placeholder="Password"
                                        onChange={formik.handleChange}
                                        onBlur={formik.handleBlur}
                                        value={formik.values.password}
                                        isInvalid={formik.touched.password && !!formik.errors.password}
                                    />
                                    <Form.Control.Feedback type="invalid">{formik.errors.password}</Form.Control.Feedback>
                                </FloatingLabel>
                            </Col>
                        </Row>

                        { checkUserPass ?
                        <Row>
                            <Col>
                                <Alert variant="danger">Invalid username/password</Alert>
                            </Col>
                        </Row>
                        : null }

                        <Row>
                            <Col>
                            <Button variant="primary" type="submit" size="sm" disabled={ btnSubmitLoading }>
                                { btnSubmitLoading ? <Spinner as="span" animation="border" size="sm" role="status" aria-hidden="true" /> : "Login!" }
                            </Button>
                            </Col>
                        </Row>
                    </Form>
                </Card>
            </div>
        </div>
    );
}

export default LoginForm;