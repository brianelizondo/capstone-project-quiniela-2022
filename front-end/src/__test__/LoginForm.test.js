import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
import { persistStore } from 'redux-persist';
import store from '../store/store';
import LoginForm from '../LoginForm';
let persistor = persistStore(store);

// Smoke Test
describe("Smoke Test", () => {
    it("renders 'LoginForm' component without crashing", () => {
        const userAuthenticate = jest.fn();
        render(
            <Provider store={store}>
                <PersistGate loading={null} persistor={persistor}>
                    <LoginForm userAuthenticate={userAuthenticate} />
                </PersistGate>
            </Provider>
        );
        const textFound = screen.getByText("Login to Quinielas World Cup 2022");
        expect(textFound).toBeInTheDocument();
    });
});

// Snapshot Tests
describe("Snapshot Test", () => {
    it("'LoginForm' matches snapshot", () => {
        const userAuthenticate = jest.fn();
        const {asFragment} = render(
            <Provider store={store}>
                <PersistGate loading={null} persistor={persistor}>
                    <LoginForm userAuthenticate={userAuthenticate} />
                </PersistGate>
            </Provider>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});

describe("Additional Test", () => {
    it("should show a required field warning for each empty input field", async () => {
        const userAuthenticate = jest.fn();
        render(
            <Provider store={store}>
                <PersistGate loading={null} persistor={persistor}>
                    <LoginForm userAuthenticate={userAuthenticate} />
                </PersistGate>
            </Provider>
        );
        const user = userEvent.setup();
    
        await user.click(screen.getByRole("button", {type: /Login!/i}));
        const warnings = await screen.findAllByText("Required");
        expect(warnings[0]).toBeVisible();
    });

    it("should show invalid field errors for each invalid input field", async () => {
        const userAuthenticate = jest.fn();
        render(
            <Provider store={store}>
                <PersistGate loading={null} persistor={persistor}>
                    <LoginForm userAuthenticate={userAuthenticate} />
                </PersistGate>
            </Provider>
        );
        const user = userEvent.setup();

        await user.type(screen.getByLabelText(/username/i), 'test');
        await user.type(screen.getByLabelText(/password/i), '12345')
        await user.click(screen.getByRole("button", {type: /Login!/i}));
        let warnings = await screen.findAllByText("Must be 8 characters or more");
        expect(warnings[0]).toBeVisible();

        await user.type(screen.getByLabelText(/username/i), 'testuser..');
        await user.type(screen.getByLabelText(/password/i), '12345..')
        await user.click(screen.getByRole("button", {type: /Login!/i}));
        warnings = await screen.findAllByText("Only letters and numbers are allowed");
        expect(warnings[0]).toBeVisible();
    });
});