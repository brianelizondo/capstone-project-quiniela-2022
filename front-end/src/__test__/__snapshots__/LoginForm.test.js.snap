// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test 'LoginForm' matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="RegisterForm col-md-8 offset-md-2"
  >
    <h1>
      Login
    </h1>
    <p>
      Login to Quinielas World Cup 2022
    </p>
    <div
      class="col-md-8 offset-md-2"
    >
      <form
        class=""
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="mb-3 form-floating"
            >
              <input
                class="form-control"
                id="username"
                name="username"
                placeholder="Username"
                type="text"
                value=""
              />
              <div
                class="invalid-feedback"
              />
              <label
                for="username"
              >
                Username
              </label>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="mb-3 form-floating"
            >
              <input
                class="form-control"
                id="password"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
              <div
                class="invalid-feedback"
              />
              <label
                for="password"
              >
                Password
              </label>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          >
            <button
              class="btn btn-primary btn-sm"
              type="submit"
            >
              Login  !
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
